crates/wasm-mutate/src/mutators/modify_data.rs

```diff
@@ -0,0 +1,113 @@
+        let data_to_modify = config.rng().gen_range(0, reader.get_count());
+
+        // Iterate over all data segments in the old data section and re-add
+        // them to the `new_section` one-by-one.
+        for i in 0..reader.get_count() {
+            let data = reader.read()?;
+            let offset;
+            // Preserve the mode of the data segment
+            let mode = match &data.kind {
+                DataKind::Active {
+                    memory_index,
+                    init_expr,
+                } => {
+                    offset = DefaultTranslator.translate_init_expr(init_expr)?;
+                    DataSegmentMode::Active {
+                        memory_index: *memory_index,
+                        offset: &offset,
+                    }
+                }
+                DataKind::Passive => DataSegmentMode::Passive,
+            };
+            // If this is the correct data segment apply the mutation,
+            // otherwise preserve the data.
+            let data = if i == data_to_modify {
+                self.mutate_data(config, data.data)
+            } else {
+                data.data.to_vec()
+            };
+            new_section.segment(DataSegment { mode, data });
+        }
+
+        Ok(Box::new(std::iter::once(Ok(config
+            .info()
+            .replace_section(
+                config.info().data.unwrap(),
+                &new_section,
+            )))))
+    }
+
+    fn can_mutate<'a>(&self, config: &'a WasmMutate) -> bool {
+        // Modifying a data segment doesn't preserve the semantics of the
+        // original module and also only works if there's actually some data.
+        !config.preserve_semantics && config.info().num_data() > 0
+    }
+}
+
+impl ModifyDataMutator {
+    fn mutate_data(&self, config: &mut WasmMutate, data: &[u8]) -> Vec<u8> {
+        // If reduction is configured then we never add data, otherwise
+        // arbitrarily choose to add or remove data.
+        if config.reduce || config.rng().gen() {
+            if data.len() > 0 {
+                // Remove a random subslice of data, if there's actually some
+                // data.
+                let start = config.rng().gen_range(0, data.len());
+                let end = config.rng().gen_range(start, data.len());
+                data[start..end].to_vec()
+            } else {
+                Vec::new()
+            }
+        } else {
+            // Insert some random bytes.
+            let start = config.rng().gen_range(0, data.len());
+            let len = config.rng().gen_range(0, 100);
+            data[0..start]
+                .iter()
+                .copied()
+                .chain(
+                    config
+                        .rng()
+                        .sample_iter(rand::distributions::Standard)
+                        .take(len),
+                )
+                .chain(data[start..].iter().copied())
+                .collect()
+        }
+    }
```

Comment:
I think we should (either in a third logic branch or maybe replacing both branches but configuring the max sizes appropriately) add a mode where we call the configured mutate-this-raw-data hook which we can make call `libfuzzer`'s internal mutators directly when we are fuzzing.

https://github.com/bytecodealliance/wasm-tools/blob/5d4f00722c4ed496336ef1dd9a7742a1a3b171f9/crates/wasm-mutate/src/lib.rs#L233-L246

https://docs.rs/libfuzzer-sys/latest/libfuzzer_sys/fn.fuzzer_mutate.html

Reply:
To make sure I understand, you're imagining this saw `raw_mutate_func`-style hook being configured in `WasmMutate`? Something like an optional closure where, if present, is called here?

Reply:
Yeah I think ideally we would write our own, default `raw_mutate_func` implementation and then just always call `raw_mutate_func` here, instead of maintaining N different code paths that do shrinking vs not shrinking etc. As long as `raw_mutate_func` abides by the size constraints passed to it, we shouldn't need separate branches for adding vs deleting vs editing.

Reply:
How would that work though with the seed? We wouldn't pass in the seed or the `Unstructured` so what would the consumer used as the basis for mutating the input data?

Also if you're imagining that this is always an option should the CLI be the only one that implements this? By default would this mutator be disabled unless the callback were configured?

(sorry I'm not really sure how you're envisioning this working)

Reply:
Hm yeah we should probably pass a seed into this callback.

We generally aren't using `Unstructured` here, not sure what you're getting at here.

In practice, we will hook `raw_mutate_func` up to https://docs.rs/libfuzzer-sys/latest/libfuzzer_sys/fn.fuzzer_mutate.html

If the `raw_mutate_func` is not configured, then we should use a default, internal implementation of the `raw_mutate_func` callback, similar to what you've done here but not hard-coded to be only for data segments.

```rust
impl WasmMutate {
    fn get_raw_mutate_func(&self) -> &dyn Fn(&mut Vec<u8>, ...) {
        self.raw_mutate_func.as_ref().unwrap_or(&default_raw_mutate_func)
    }
}

fn default_raw_mutate_func(
    data: &mut Vec<u8>, 
    // probably need to enrichen the `raw_mutate_func` signature with
    // a seed and info about whether we are only shrinking...
) {
    // ... the stuff you are doing for data segments right now ...
}

// ...

// Inside the data segment mutator.
(config.raw_mutate_func())(&mut my_data_segment);
```

Does that make sense?

> Also if you're imagining that this is always an option should the CLI be the only one that implements this? By default would this mutator be disabled unless the callback were configured?

I would say the other way around: the CLI shouldn't expose this functionality because what different kinds of implementations can it have? We're probably not going to pipe out to another program to mutate these little bits of data, so there isn't really much to configure other than maybe the source of entropy. Not really useful here.

However, when using `wasm-mutate` as a library inside a fuzz target, we definitely want to hook this up to https://docs.rs/libfuzzer-sys/latest/libfuzzer_sys/fn.fuzzer_mutate.html so that we can let libfuzzer use its dictionary and all of its builtin mutation schemes when mutating these data segments or export names or what have you.

Reply:
Sorry I was mixing up sources of randomness  and got confused with `Unstructured` vs just a plain old `Rng` here.

Also thanks for explaining! I've pushed up something which I think roughly does what you're thinking, but I think it'd be good to double-check as well.